o1vm/src/pickles/main.rs

```diff
@@ -158,38 +159,87 @@ pub fn cannon_main(args: cli::cannon::RunArgs) {
-            debug!(
-                "Proof generated in {elapsed} μs",
-                elapsed = start_iteration.elapsed().as_micros()
-            );
-            {
-                let start_iteration = Instant::now();
-                let verif = verifier::verify::<
-                    Vesta,
-                    DefaultFqSponge<VestaParameters, PlonkSpongeConstantsKimchi>,
-                    DefaultFrSponge<Fp, PlonkSpongeConstantsKimchi>,
-                >(domain_fp, &srs, &constraints, &proof);
-                debug!(
-                    "Verification done in {elapsed} μs",
-                    elapsed = start_iteration.elapsed().as_micros()
-                );
-                assert!(verif);
-            }
+            prove_and_verify(domain_fp, &srs, &constraints, curr_proof_inputs, &mut rng);
 
             curr_proof_inputs = ProofInputs::new(domain_size);
         }
     }
+
+    if curr_proof_inputs.evaluations.instruction_counter.len() < domain_size {
+        debug!("Padding witness for proof generation");
+        pad(&mips_wit_env, &mut curr_proof_inputs, &mut rng);
+        prove_and_verify(domain_fp, &srs, &constraints, curr_proof_inputs, &mut rng);
+    }
+}
+
+fn prove_and_verify(
+    domain_fp: EvaluationDomains<Fp>,
+    srs: &SRS<Vesta>,
+    constraints: &[E<Fp>],
+    curr_proof_inputs: ProofInputs<Vesta>,
+    rng: &mut ThreadRng,
+) {
+    let start_iteration = Instant::now();
+    let proof = prover::prove::<
+        Vesta,
+        DefaultFqSponge<VestaParameters, PlonkSpongeConstantsKimchi>,
+        DefaultFrSponge<Fp, PlonkSpongeConstantsKimchi>,
+        _,
+    >(domain_fp, srs, curr_proof_inputs, constraints, rng)
+    .unwrap();
+    debug!(
+        "Proof generated in {elapsed} μs",
+        elapsed = start_iteration.elapsed().as_micros()
+    );
+    let start_iteration = Instant::now();
+    let verif = verifier::verify::<
+        Vesta,
+        DefaultFqSponge<VestaParameters, PlonkSpongeConstantsKimchi>,
+        DefaultFrSponge<Fp, PlonkSpongeConstantsKimchi>,
+    >(domain_fp, srs, constraints, &proof);
+    debug!(
+        "Verification done in {elapsed} μs",
+        elapsed = start_iteration.elapsed().as_micros()
+    );
+    assert!(verif);
+}
+
+fn pad(
+    witness_env: &mips_witness::Env<Fp, Box<dyn PreImageOracleT>>,
+    curr_proof_inputs: &mut ProofInputs<Vesta>,
+    rng: &mut ThreadRng,
+) {
+    let zero = Fp::zero();
+    // FIXME: Find a better way to get instruction selectors that doesn't
+    // reveal internals.
+    let noop_selector: Fp = {
+        let noop: usize = Instruction::NoOp.into();
+        Fp::from((noop - N_MIPS_REL_COLS) as u64)
+    };
+    curr_proof_inputs
+        .evaluations
+        .scratch
+        .iter_mut()
+        .for_each(|x| x.resize(x.capacity(), zero));
+    curr_proof_inputs
+        .evaluations
+        .scratch_inverse
+        .iter_mut()
+        .for_each(|x| x.resize(x.capacity(), zero));
```

Comment:
Even though we use `Vec::with_capacity(domain_size)`, I would be in favor of using the domain size, as this piece of code, even though we manage it in the same codebase, is not supposed to know how the structure `ProofInputs` is initialized.